### ISSUE TITLE
Setting static arguments in `torch.compile()`

### ISSUE THREAD

**Issue by keunhong** [pytorch/pytorch#121299]
### 🚀 The feature, motivation and pitch

JAX's `jax.jit` function allows you to set `static_argnums` or `static_argnames`. These arguments will be compiled into the function.

https://jax.readthedocs.io/en/latest/_autosummary/jax.jit.html

There doesn't seem to be a way to do this in PyTorch.

**Comment by ezyang** [comment]
What kind of argument types are you thinking of? If they are int/str, they are *already* specialized by default haha (and you have to work to unspecialize something like int)

**Comment by cagrikymk** [comment]
I am also curious if there is any functionality that corresponds to jax's static arguments.

In my case, I need a static control variable that decides whether to compute some extra term or not. If not set, the function is supposed to return None as the extra term.

So, basically:
```
@torch.jit.script
def my_func1(calc_extra):
    val1 = compute1()
    if calc_extra:
        val2 = compute2(val1)
    else:
        val2 = None
    return val1, val2

```

```
@torch.jit.script
def my_func2():
    calc_extra = True
    val1 = compute1()
    if calc_extra:
        val2 = compute2(val1)
    return val1, val2

```

While this code works, my_func1(True) is slower than my_func2()  (For my use-case the difference is around 30-40%).
I assume the performance takes a hit because of the control logic which doesnt happen for JAX as when the control variable defined as static.

**Comment by ezyang** [comment]
Have you tried it with torch.compile? I expect these to have equivalent performance in torch.compile. In fact, it is very hard to get torch.compile to not make things static unless they are tensors. Tensor static arguments is a legitimate feature request but it's more niche.

**Comment by cagrikymk** [comment]
@ezyang The performance difference disappears when I use torch.compile but torch.jit.script works significantly better for my code. I assume this is out of context for this discussion. 

Thank you!

**Comment by ezyang** [comment]
Yeah, you are welcome to file a bug for static arguments in TS but you aren't likely to get help for it
